csharp/test/Apache.Arrow.IntegrationTest/CDataInterface.cs

```diff
@@ -0,0 +1,79 @@
+// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Data;
+using System.Diagnostics;
+using System.IO;
+using Apache.Arrow.C;
+using Apache.Arrow.Arrays;
+using Apache.Arrow.Types;
+
+namespace Apache.Arrow.IntegrationTest
+{
+    /// <summary>
+    /// Bridge for C Data Interface integration testing.
+    /// These methods are called from the Python integration testing
+    /// harness provided by Archery.
+    /// </summary>
+    public static class CDataInterface
+    {
+        // Archery uses the `pythonnet` library (*) to invoke .Net DLLs.
+        // `pythonnet` is only able to marshal simple types such as int and
+        // str, which is why we provide trivial wrappers around other APIs.
+        //
+        // (*) https://pythonnet.github.io/
+
+        public static void Initialize()
+        {
+            // Allow debugging using Debug.WriteLine()
+            Trace.Listeners.Add(new ConsoleTraceListener());
+        }
+
+        public static unsafe Schema ImportSchema(long ptr)
+        {
+            return CArrowSchemaImporter.ImportSchema((CArrowSchema*) ptr);
+        }
+
+        public static unsafe void ExportSchema(Schema schema, long ptr)
+        {
+            CArrowSchemaExporter.ExportSchema(schema, (CArrowSchema*) ptr);
+        }
+
+        public static unsafe RecordBatch ImportRecordBatch(long ptr, Schema schema)
+        {
+            return CArrowArrayImporter.ImportRecordBatch((CArrowArray*) ptr, schema);
+        }
+
+        public static unsafe void ExportRecordBatch(RecordBatch batch, long ptr)
+        {
+            CArrowArrayExporter.ExportRecordBatch(batch, (CArrowArray*) ptr);
+        }
+
+        public static JsonFile ParseJsonFile(String jsonPath)
+        {
+            return JsonFile.Parse(new FileInfo(jsonPath));
+        }
+
+        public static long GetAllocatedBytes()
+        {
+            GC.Collect();
+            // XXX this doesn't seem to give stable and reliable measurements
+            var gcInfo = GC.GetGCMemoryInfo();
+            return gcInfo.PromotedBytes;
```

Comment:
I wasn't able to find a formula that produced reliable numbers. Inevitably, the numbers are fluctuating (going up _and_ down). I don't know if the .Net GC is just giving us approximate numbers, or if some internal caches are interfering.

Reply:
Are we sure that all the tests are using the TestMemoryAllocator (which uses the managed heap)? I suspect some arrays may be allocated using the NativeMemoryAllocator which does not used the managed heap.

Whether or not it's strictly required, I've developed a habit of something like

for (int i = 0; i < 3; i++) {
    GC.Collect();
    GC.WaitForPendingFinalizers();
}

and this seems to sometimes make a difference.

Reply:
I also don't know what kind of effect pythonnet would have on this. There may be Python objects waiting for a Python GC that are keeping .NET objects alive.

Reply:
> Are we sure that all the tests are using the TestMemoryAllocator (which uses the managed heap)? I suspect some arrays may be allocated using the NativeMemoryAllocator which does not used the managed heap.

Hmm, I'm discovering the C# codebase, but this is what I get:
1. the JSON reader in Apache.Arrow.IntegrationTest just uses the default memory allocator when calling `ArrowBuffer.Builder.Build()`
2. the TestMemoryAllocator doesn't implement `IOwnableAllocation`, which is required by `CArrowArrayExporter` for all exported buffers

In any case, however, the problem I'm mentioning in the comment above (unstable measurements) can be observed even on the Schema tests, which shouldn't invoke an external allocator.

> I also don't know what kind of effect pythonnet would have on this. There may be Python objects waiting for a Python GC that are keeping .NET objects alive.

That would seem surprising, but I can try a `gc.collect()` on the Python side as well...

Reply:
Okay, none of your suggestions works unfortunately.

Reply:
My knowledge of the GC is less like science and more like voodoo. I'm also not familiar with this API. That said, PromotedBytes seems like it's probably not the right choice here and I would expect HeapSizeBytes to be the value that's actually wanted.

Reply:
Ah, but I also tried `GC.GetTotalMemory()`, `HeapSizeBytes` and also `HeapSizeBytes - FragmentedBytes`. None gave stable values.